File: src/main/java/com/tvl/sql/Lexer.java
package com.tvl.sql;

import java.util.ArrayList;
import java.util.List;

/**
 * 极简词法分析器。Token 类型：
 *  关键字/标识符（IDENT）、数字（NUMBER）、单引号字符串（STRING）、
 *  参数 ?（PARAM）、比较/括号/逗号/星号等符号。
 * 关键字与标识符的区分在解析器里按字面做（大小写不敏感）。
 */
final class Lexer {

    enum Kind {
        IDENT, NUMBER, STRING, PARAM,
        LPAREN, RPAREN, COMMA, STAR,
        OP, // = <> != < <= > >=
        EOF
    }

    static final class Token {
        final Kind kind;
        final String text;
        final int pos;

        Token(Kind kind, String text, int pos) {
            this.kind = kind;
            this.text = text;
            this.pos = pos;
        }

        @Override
        public String toString() {
            return kind + "(" + text + ")@" + pos;
        }
    }

    private final String src;
    private int pos;

    private Lexer(String src) {
        this.src = src;
    }

    static List<Token> tokenize(String sql) {
        return new Lexer(sql).run();
    }

    private List<Token> run() {
        List<Token> tokens = new ArrayList<>();
        while (true) {
            skipWhitespaceAndComments();
            if (pos >= src.length()) {
                tokens.add(new Token(Kind.EOF, "", pos));
                return tokens;
            }
            char c = src.charAt(pos);
            int start = pos;
            if (c == '?') {
                pos++;
                tokens.add(new Token(Kind.PARAM, "?", start));
            } else if (c == '(') {
                pos++;
                tokens.add(new Token(Kind.LPAREN, "(", start));
            } else if (c == ')') {
                pos++;
                tokens.add(new Token(Kind.RPAREN, ")", start));
            } else if (c == ',') {
                pos++;
                tokens.add(new Token(Kind.COMMA, ",", start));
            } else if (c == '*') {
                pos++;
                tokens.add(new Token(Kind.STAR, "*", start));
            } else if (c == '\'') {
                tokens.add(readString());
            } else if (Character.isDigit(c) || (c == '.' && pos + 1 < src.length()
                    && Character.isDigit(src.charAt(pos + 1)))) {
                tokens.add(readNumber());
            } else if (isIdentStart(c)) {
                tokens.add(readIdent());
            } else if (c == '=' || c == '<' || c == '>' || c == '!') {
                tokens.add(readOperator());
            } else {
                throw new SqlParseException("无法识别的字符 '" + c + "'（位置 " + pos + "）");
            }
        }
    }

    private void skipWhitespaceAndComments() {
        while (pos < src.length()) {
            char c = src.charAt(pos);
            if (Character.isWhitespace(c)) {
                pos++;
            } else if (c == '-' && pos + 1 < src.length() && src.charAt(pos + 1) == '-') {
                pos += 2;
                while (pos < src.length() && src.charAt(pos) != '\n') {
                    pos++;
                }
            } else if (c == '/' && pos + 1 < src.length() && src.charAt(pos + 1) == '*') {
                pos += 2;
                while (pos + 1 < src.length()
                        && !(src.charAt(pos) == '*' && src.charAt(pos + 1) == '/')) {
                    pos++;
                }
                if (pos + 1 >= src.length()) {
                    throw new SqlParseException("块注释没有闭合");
                }
                pos += 2;
            } else {
                return;
            }
        }
    }

    private Token readString() {
        int start = pos;
        pos++; // 开引号
        StringBuilder sb = new StringBuilder();
        while (pos < src.length()) {
            char c = src.charAt(pos);
            if (c == '\'') {
                // SQL 转义：'' -> '
                if (pos + 1 < src.length() && src.charAt(pos + 1) == '\'') {
                    sb.append('\'');
                    pos += 2;
                } else {
                    pos++;
                    return new Token(Kind.STRING, sb.toString(), start);
                }
            } else {
                sb.append(c);
                pos++;
            }
        }
        throw new SqlParseException("字符串缺少闭合单引号（位置 " + start + "）");
    }

    private Token readNumber() {
        int start = pos;
        boolean seenDot = false;
        while (pos < src.length()) {
            char c = src.charAt(pos);
            if (Character.isDigit(c)) {
                pos++;
            } else if (c == '.' && !seenDot) {
                seenDot = true;
                pos++;
            } else {
                break;
            }
        }
        return new Token(Kind.NUMBER, src.substring(start, pos), start);
    }

    private Token readIdent() {
        int start = pos;
        while (pos < src.length() && isIdentPart(src.charAt(pos))) {
            pos++;
        }
        return new Token(Kind.IDENT, src.substring(start, pos), start);
    }

    private Token readOperator() {
        int start = pos;
        char c = src.charAt(pos);
        if (c == '=') {
            pos++;
            return new Token(Kind.OP, "=", start);
        }
        if (c == '<') {
            if (pos + 1 < src.length()) {
                char n = src.charAt(pos + 1);
                if (n == '=') {
                    pos += 2;
                    return new Token(Kind.OP, "<=", start);
                }
                if (n == '>') {
                    pos += 2;
                    return new Token(Kind.OP, "<>", start);
                }
            }
            pos++;
            return new Token(Kind.OP, "<", start);
        }
        if (c == '>') {
            if (pos + 1 < src.length() && src.charAt(pos + 1) == '=') {
                pos += 2;
                return new Token(Kind.OP, ">=", start);
            }
            pos++;
            return new Token(Kind.OP, ">", start);
        }
        // c == '!'
        if (pos + 1 < src.length() && src.charAt(pos + 1) == '=') {
            pos += 2;
            return new Token(Kind.OP, "!=", start);
        }
        throw new SqlParseException("非法运算符（位置 " + pos + "，是否想写 <> 或 != ？）");
    }

    private static boolean isIdentStart(char c) {
        return Character.isLetter(c) || c == '_';
    }

    private static boolean isIdentPart(char c) {
        return Character.isLetterOrDigit(c) || c == '_';
    }
}

File: resflow/lexer.py
"""Hand-written lexer for the ResFlow language.

The lexer is deliberately dependency free: it tracks byte offsets and maps
them to 1-based line/column pairs so every token (and therefore every AST
node and diagnostic) can be traced back to the exact source span.
"""

from dataclasses import dataclass

from .errors import ResFlowError

# Reserved words.  ``use`` is reserved because USE-FREE hinges on it.
KEYWORDS = {
    "fn", "let", "acquire", "release", "use", "return",
    "throw", "if", "else", "while", "try", "catch",
}
# Literal keywords get their own token kind in _lex_word.
LITERAL_KEYWORDS = {"true", "false"}


@dataclass(frozen=True)
class Location:
    """A half-open source span ``[offset, end_offset)`` with line/column."""

    line: int
    column: int
    offset: int
    end_offset: int

    def to_dict(self):
        return {
            "line": self.line,
            "column": self.column,
            "offset": self.offset,
            "end_offset": self.end_offset,
        }


class Token:
    __slots__ = ("kind", "value", "loc")

    def __init__(self, kind, value, loc):
        self.kind = kind
        self.value = value
        self.loc = loc

    def __repr__(self):
        return f"Token({self.kind!r}, {self.value!r}, {self.loc})"


class LexError(ResFlowError):
    pass


def _is_ident_start(ch):
    return ch == "_" or ch.isalpha()


def _is_ident_part(ch):
    return ch == "_" or ch.isalnum()


class Lexer:
    def __init__(self, source):
        self.source = source
        self.n = len(source)
        self.pos = 0
        self.line = 1
        self.line_start = 0  # offset where the current line begins

    # -- location helpers --------------------------------------------------

    def _loc(self, start, end=None):
        if end is None:
            end = start + 1
        col = start - self.line_start + 1
        return Location(self.line, col, start, end)

    def _advance(self):
        ch = self.source[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.line_start = self.pos
        return ch

    # -- public entry point ------------------------------------------------

    def tokenize(self):
        tokens = []
        src = self.source
        while self.pos < self.n:
            ch = src[self.pos]
            if ch in " \t\r\n":
                self._advance()
                continue
            if ch == "/" and self.pos + 1 < self.n and src[self.pos + 1] == "/":
                while self.pos < self.n and src[self.pos] != "\n":
                    self.pos += 1
                continue
            if ch == "/" and self.pos + 1 < self.n and src[self.pos + 1] == "*":
                self._lex_block_comment()
                continue
            start = self.pos
            if _is_ident_start(ch):
                tokens.append(self._lex_word(start))
                continue
            if ch.isdigit():
                tokens.append(self._lex_number(start))
                continue
            if ch == '"':
                tokens.append(self._lex_string(start))
                continue
            tokens.append(self._lex_punct(start))
        tokens.append(Token("EOF", None, self._loc(self.pos, self.pos)))
        return tokens

    # -- individual token kinds -------------------------------------------

    def _lex_block_comment(self):
        start = self.pos
        self._advance()  # /
        self._advance()  # *
        depth = 1
        while self.pos < self.n and depth:
            ch = self.source[self.pos]
            if ch == "/" and self.pos + 1 < self.n and self.source[self.pos + 1] == "*":
                self._advance()
                self._advance()
                depth += 1
            elif ch == "*" and self.pos + 1 < self.n and self.source[self.pos + 1] == "/":
                self._advance()
                self._advance()
                depth -= 1
            else:
                self._advance()
        if depth:
            raise LexError("unterminated block comment", self._loc(start, self.pos))

    def _lex_word(self, start):
        self._advance()
        while self.pos < self.n and _is_ident_part(self.source[self.pos]):
            self._advance()
        text = self.source[start:self.pos]
        if text in LITERAL_KEYWORDS:
            return Token("BOOL", text, self._loc(start, self.pos))
        if text in KEYWORDS:
            return Token("KEYWORD", text, self._loc(start, self.pos))
        return Token("IDENT", text, self._loc(start, self.pos))

    def _lex_number(self, start):
        self._advance()
        while self.pos < self.n and self.source[self.pos].isdigit():
            self._advance()
        text = self.source[start:self.pos]
        return Token("INT", int(text), self._loc(start, self.pos))

    def _lex_string(self, start):
        self._advance()  # opening quote
        buf = []
        while self.pos < self.n and self.source[self.pos] != '"':
            ch = self.source[self.pos]
            if ch == "\\":
                self._advance()
                if self.pos >= self.n:
                    break
                esc = self.source[self.pos]
                buf.append({"n": "\n", "t": "\t", '"': '"',
                            "\\": "\\", "0": "\0"}.get(esc, esc))
                self._advance()
            elif ch == "\n":
                raise LexError("unterminated string literal",
                               self._loc(start, self.pos))
            else:
                buf.append(ch)
                self._advance()
        if self.pos >= self.n:
            raise LexError("unterminated string literal",
                           self._loc(start, self.pos))
        self._advance()  # closing quote
        return Token("STRING", "".join(buf), self._loc(start, self.pos))

    # Two-character operators are matched greedily before single characters.
    _PUNCT2 = {"==", "!=", "<=", ">=", "&&", "||"}
    _PUNCT1 = set("(){};,=!<>&|+-*/%")

    def _lex_punct(self, start):
        two = self.source[self.pos:self.pos + 2]
        if two in self._PUNCT2:
            self._advance()
            self._advance()
            # Multi-character operators carry their spelling as the kind.
            return Token(two, two, self._loc(start, self.pos))
        ch = self.source[self.pos]
        if ch in self._PUNCT1:
            self._advance()
            # Single-character punctuation uses the canonical kind names
            # the parser expects; the spelling is kept in ``value``.
            kind = {"(": "LPAREN", ")": "RPAREN", "{": "LBRACE",
                    "}": "RBRACE", ";": "SEMI", ",": "COMMA"}.get(ch, ch)
            return Token(kind, ch, self._loc(start, self.pos))
        raise LexError(f"unexpected character {ch!r}", self._loc(start))
